src/utils/changeChain.ts

```diff
@@ -0,0 +1,51 @@
+import {
+  internalAddressesPerNetwork,
+  ValidNetworkNameTypes,
+} from "../utils/contracts/contractAddresses/internalAddresses";
+import { injectedConnector } from "../hooks/injectedConnectors";
+
+function switchEthereumChain(connector: any, chainId: number) {
+  connector.request({
+    id: 1,
+    jsonrpc: "2.0",
+    method: "wallet_switchEthereumChain",
+    params: [
+      {
+        chainId: "0x" + chainId.toString(16),
+      },
+    ],
+  });
+}
+function addEthereumChain(connector: any, chain: any) {
+  connector.request({
+    id: 1,
+    jsonrpc: "2.0",
+    method: "wallet_addEthereumChain",
+    params: [
+      {
+        chainId: "0x" + chain.chainId.toString(16),
+        chainName: chain.chainName,
+        rpcUrls: [chain.rpcUrl],
+        nativeCurrency: {
+          name: chain.symbol,
+          symbol: chain.symbol,
+          decimals: 18,
+        },
+        blockExplorerUrls: [chain.explorer],
+      },
+    ],
+  });
+}
+
+// Only supported with Metamask
+export function changeChain(chainName: ValidNetworkNameTypes) {
+  const chain = internalAddressesPerNetwork[chainName];
+
+  injectedConnector.getProvider().then((connector: any) => {
+    if (chain.chainId === 4) {
+      switchEthereumChain(connector, chain.chainId);
+    } else {
+      addEthereumChain(connector, chain);
+    }
+  });
```

Comment:
Instead of throwing this out based on a guess of their configuration, then firing and forgetting the RPC call- why not try awaiting the result of the `switch` call, and- if it returns an error which states that the specified chain is not registered- calls and awaits `addEthereumChain` and switches to it if successful?

Reply:
I'm thinking on something like this:
```
  injectedConnector.getProvider().then(async (provider: any) => {
    try {
      await switchEthereumChain(provider, chain.chainId);
    } catch {
      await addEthereumChain(provider, chain);
    }
  });
```

Reply:
I'm making this change to this 
```
injectedConnector.getProvider().then(async (provider: any) => {
  await provider
    .request(createRequestArguments("wallet_switchEthereumChain", chain))
    .catch(
      provider.request(
        createRequestArguments("wallet_addEthereumChain", chain)
      )
    );
});
```

Reply:
Looking at the details, it appears MetaMask is a bit iffy with how it handles this sort of thing. I'll clean it up a bit.